Improved error messages and UI when importing GPS trace files

nenad-vujicic opened the pull request:
PR adds client and server side checking if supplied trace file has GPX extension

Adds "GPX File (*.gpx)" filter to Open File dialog. Adds checking to TracesController#create if sent file has "gpx" extension (before parsing) by first converting file name to lowercase and then checking if it ends with .gpx. In case of failure, it reports error using bootstrap alert:

![image](https://github.com/user-attachments/assets/5b786adf-1351-4064-b5a2-45ea4a0dacf0)

mmd-osm commented:
Please note that we’re also accepting zip files (and other compressed formats) with gpx files inside.

tomhughes commented:
As @mmd-osm says there is no such restriction - even ignoring the fact that we accept a wide variety of formats that are not simple GPX files it's not clear to me that there's any good reason to insist on a `.gpx` extension even for GPX files.

nenad-vujicic commented:
I apologize for this mistake .. I forgot the archives types. Just to be sure, are these file formats (I collected list from Trace.rb: ".tar.gz", ".tar.bz2", ".tar", ".zip", ".gpx.gz", ".gpx.bz2", ".gpx") the only supported or there are more (.json, .csv, ..)?

Thanks.

AntonKhorev commented:
The problem in https://github.com/openstreetmap/openstreetmap-website/issues/2630 was that the email called the uploaded file "your GPX file" even when it wasn't a gpx file and then listed parse error messages. Unexpected file extensions wasn't the problem.

gravitystorm commented:
> I apologize for this mistake .. I forgot the archives types. Just to be sure, are these file formats (I collected list from Trace.rb: ".tar.gz", ".tar.bz2", ".tar", ".zip", ".gpx.gz", ".gpx.bz2", ".gpx") the only supported or there are more (.json, .csv, ..)?

We don't use filenames to detect what type of file has been uploaded. Extensions are optional and shouldn't be used for filtering. If you have a file called "some_trace" then we will detect if it is an archive (using `/usr/bin/file`, a compressed archive, or similar, and handle it appropriately.

The `extension_name` is used when you are later downloading a trace that was successfully imported - we will synthesis a name for the file based on the trace id, and then pick an appropriate extension based on the format of the file, again using `/usr/bin/file` for detection.

We can certainly improve the error handling for when someone uploads a file that can't be parsed!

nenad-vujicic commented:
Here is the latest version:

GPS Trace upload page:
![image](https://github.com/user-attachments/assets/4e5f07c3-fae1-4471-aeff-831e329cee76)
HTML email after importing failure:
![image](https://github.com/user-attachments/assets/54d8250b-d1c2-4888-8b15-8b577147de20)

1. Replaced "Upload GPX File" with "Choose GPS Trace File"
2. Added supported files extensions to upload page
3. Removed GPX from "It looks like your GPX file"
4. Improved import failure message sent by email

Please, let me know your thoughts.

Thanks!

nenad-vujicic commented:
@mmd-osm , do you think the latest solution could solve the issue #2630 you reported? Any suggestion how to improve it?

Thanks!

AntonKhorev commented on app/views/traces/new.html.erb:
```diff
@@ -3,7 +3,8 @@
 <%= bootstrap_form_for @trace, :url => { :action => "create" }, :html => { :multipart => true } do |f| %>
-  <%= f.file_field :gpx_file, :placeholder => t("helpers.file.prompt") %>
+  <%= f.file_field :gpx_file, :placeholder => t("helpers.file.prompt"),
+                              :help => ".gpx, .tar.gz, .tar.bz2, .tar, .zip, .gpx.gz, .gpx.bz2" %>
```
Are you not putting the help string into `en.yml` [help scope](https://github.com/openstreetmap/openstreetmap-website/blob/0423d18a8f5b232aa2a08b55351bee553047fb81/config/locales/en.yml#L153) on purpose, for it not to show up in translations because there's nothing to translate?

nenad-vujicic replied:
Yes, I couldn't imagine how I could translate list of file formats. But, if you think we should put it, I can put it?
